src/edit.rs

```diff
@@ -625,13 +652,28 @@ impl<'out, 'prompt, H: Helper> State<'out, 'prompt, H> {
         ) {
             self.ctx.history_index = history_index;
             let buf = history.get(history_index).unwrap();
-            self.changes.borrow_mut().begin();
-            self.line.update(buf, buf.len());
-            self.changes.borrow_mut().end();
-            self.refresh_line()
-        } else {
-            self.out.beep()
+
+            // if the history search results in the same command we found last, continue searching
+            // for a different command
+            if buf == self.line.as_str() {
+                return self.edit_history_search(dir, keep_cursor_pos);
+            }
+
+            let pos = if keep_cursor_pos {
+                self.line.pos()
+            } else {
+                buf.len()
+            };
+            self.update_line(buf, pos);
+        } else {
+            self.ctx.history_index = match dir {
+                Direction::Forward => history.len(),
+                Direction::Reverse => 0,
+            };
+            self.restore();
```

Comment:
bash `history-search-backward` does not seem to loop or skip same history entries.
But you are right: cursor is preserved.